Fast boot 1218 working

wudping opened the pull request:
Modularize the libyami code, so that ecoder, decoder, vpp, can be compiled separately with the existing compilation options. It helps to reduce the size of the dynamic or static library.

wudping commented:
please help to review, thanks.

uartie commented on interface/Yami.h:
```diff
@@ -18,11 +18,18 @@
 #include <YamiVersion.h>
+#include <VideoCommonDefs.h>
 
+#if (__BUILD_DECODER__)
```
It is not a good idea to do this in a public header.  It forces clients to define these also.  Also, if client project defines `__BUILD_DECODER__` (etc...) and libyami was compiled without `__BUILD_DECODER__` (etc...), then problems occur.  Thus, it's better not to impose these complexities onto clients.

Perhaps you could generate Yami.h at libyami compile time to insert appropriate includes instead.

wudping replied:
Thanks for your review and helpful suggestion @uartie ;
But here one more explaination about this piece of code, the macros like __BUILD_DECODER__ are defined in the file YamiVersion.h, clients not have to define them, it can be opened or closed according to the compilation options.
Yes, I agree with you about your suggestion, generation Yami.h is better.

uartie replied:
Ah yes, I missed those defines in YamiVersion.h.  Thanks for pointing them out.  In that case, the solution is okay for me.  The only thing I would change are the names of the defines; `__BUILD_DECODER__` (etc...) are too generic.  See my suggestion in YamiVersion.h comment.

uartie replied:
Of course, if generating Yami.h... then you could:
```
#if @__BUILD_DECODER__@
#include <VideoDecoderHost.h>
#endif
```
...to eliminate the `#defines` altogether.

uartie commented on interface/YamiVersion.h.in:
```diff
@@ -22,6 +22,15 @@
+ * __BUILD_VPP__: represent wheter vpp subdirectory is compiled or not.
+ *
+ */
+#define __BUILD_DECODER__    @__BUILD_DECODER__@
```
Perhaps something like `#define __YAMI_HAS_DECODER__ @__BUILD_DECODER__@` (etc.) would be less generic and reduce the chance of name clashes in client projects.  Also, I think these should go in Yami.h.in, instead, and then generate Yami.h since they aren't really related to versioning.

wudping commented:
hi @uartie , it's updated, thanks for your thorough information.

xuguangxin commented on interface/Yami.h.in:
```diff
@@ -19,10 +19,17 @@
 
 #include <YamiVersion.h>
 
+#if (@__YAMI_HAS_DECODER__@)
```
why we need this, include header do not enlarger the binary size

wudping replied:
agree with you @xuguangxin.  
But if no decoder is compiled, the sub-directory "decoder" will not be included into the compilation, and the base class files(vaapidecpicture.cpp, vaapidecsurfacepool.cpp...) are not be compiled,  the corresponding head files(VideoDecoderInterface.h, VideoDecoderHost.h) also are not installed, so Yami.h can not include decoder's head file(VideoDecoderHost.h).

It's okay just to install decoder's head files, and not compile the cpp files, but it will bring some confusing, and we have to change the Makefile.am too.

xuguangxin commented on decoder/vaapidecoder_base.cpp:
```diff
@@ -223,7 +224,10 @@ VaapiDecoderBase::setupVA(uint32_t numSurface, VAProfile profile)
         //use internal allocator
-        m_allocator.reset(new VaapiSurfaceAllocator(m_display->getID()), unrefAllocator);
+        if (m_noNeedExtraSurface)
+            m_allocator.reset(new VaapiSurfaceAllocator(m_display->getID(), VaapiSurfaceAllocator::MINIMUM_EXTRA_BUFFER_SIZE), unrefAllocator);
```
instead of use MINIMUM_EXTRA_BUFFER_SIZE, could we just let user set the extra buffer size?

wudping replied:
If so, user can not know how many surfaces are created.

xuguangxin commented on decoder/vaapidecoder_h264.cpp:
```diff
@@ -1625,7 +1633,11 @@ bool VaapiDecoderH264::isDecodeContextChanged(const SharedPtr<SPS>& sps)
     uint32_t height = sps->frame_cropping_flag ? sps->m_cropRectHeight
                                                : sps->m_height;
-    if (setFormat(width, height, sps->m_width, sps->m_height, maxDecFrameBuffering + 1)) {
+    if (m_configBuffer.flag & HAS_SURFACE_NUMBER && m_configBuffer.surfaceNumber > 0)
```
no need change this, the surface calculator has done by the setFormat

wudping replied:
setFormat() just use the surface number passed from format parameter.

xuguangxin commented on vaapi/vaapisurfaceallocator.cpp:
```diff
@@ -26,8 +26,8 @@
 VaapiSurfaceAllocator::VaapiSurfaceAllocator(VADisplay display, uint32_t extraSize)
-    :m_display(display),
-    m_extraSize(extraSize)
+    : m_display(display)
```
why we need change this file?

wudping replied:
it is changed by clang, because I add one NULL line.

xuguangxin commented on vaapi/vaapisurfaceallocator.h:
```diff
@@ -24,12 +24,16 @@ namespace YamiMediaCodec{
+public:
     //extra buffer size for performance
     static const uint32_t EXTRA_BUFFER_SIZE = 5;
+    static const uint32_t MINIMUM_EXTRA_BUFFER_SIZE = 2;
```
as I mentioned, this not needed. Only thing we need is let user set the extras buffer size

wudping replied:
If so, user can not know how many surfaces are created.

xuguangxin commented on codecparsers/Makefile.am:
```diff
@@ -1,9 +1,24 @@
+
+if !BUILD_FASTBOOT
+libyami_codecparser_source_c += \
+	dboolhuff.c \
```
this is a part of vp8 decoder

wudping replied:
The pass-rate of vp8 decoding is not dropped, when this file is not compiled. Maybe we can delete it.

dongpingx-wu committed:
codeparsers: delete unused files

remove files: dboolhuff.c and dboolhuff.h;

Signed-off-by: dongping wu <dongpingx.wu@intel.com>

dongpingx-wu committed:
dec: Modularize and enable to compile decoder only

It is supported to compile decoder only when encoder, vpp or some other modules
are disabled.

Signed-off-by: dongping wu <dongpingx.wu@intel.com>

dongpingx-wu committed:
enc: modularize and enable to compile encoder only

It is supported to compile encoder only when decoder, vpp or some other modules
are disabled.

Signed-off-by: dongping wu <dongpingx.wu@intel.com>

dongpingx-wu committed:
vpp: modularize and enable to compile vpp only

Enable to compile vpp module separately with the compilation option.

Signed-off-by: dongping wu <dongpingx.wu@intel.com>

dongpingx-wu committed:
v4l2: compile thread.cpp when v4l2 opened

Because thread.cpp is used by v4l2 now.

Signed-off-by: dongping wu <dongpingx.wu@intel.com>

dongpingx-wu committed:
dec: support surface number setting

Support the surface number and extra surface number setting for vp9;
low-latency and extra surface number setting for H264;

Signed-off-by: dongping wu <dongpingx.wu@intel.com>

wudping commented:
hi @xuguangxin, updated, please help to review again, thanks.

xuguangxin commented:
It's bad to change h264 dpb manager to meet fast boot requirement. We need find other way to do this
